
---
title: "Example2_predicting_multicellular_structures"
output: html_vignette
date: "2023-06-05"
vignette: >
  %\VignetteIndexEntry{Example2_predicting_multicellular_structures}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

---

```{r, echo = FALSE, message=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

In this example we'll be visualizing GEMLI results from a lineage-annotated dataset of murine intestinal crypts. Intestinal crypts originate from one or few intestinal stem cells, and are composed of a number of different cell types. In the data we find cells from individual crypts that range from just three up to forty-four related cells. The scRNA-seq dataset is derived from Bues et al. 2022 (PMID 35165449) and is publically available under Gene Expression Omnibus accession number GSE148093.


## Load package and example data

First we load the example data. Here we already predicted the lineages using GEMLI and therefore do not include a count matrix, but rather start with the predictions right away.

```{r, eval=T, echo=T}
library(GEMLI)
library(igraph)
library(HiClimR)

load('GEMLI_crypts_example_data_matrix.RData')
load('GEMLI_crypts_example_barcode_information.RData')

```

## Create a GEMLI items list

We then create a GEMLI items list. This list is used to store the data, and create and store the outputs of GEMLI (for details check example one).

```{r, eval=T, echo=T}

GEMLI_items_crypts = list()
GEMLI_items_crypts[['prediction']] = Crypts
GEMLI_items_crypts[['barcodes']] = Crypts_bc_dict

```

## Visualize predictions as network

To visualize large lineages we'll use three different network layout algorithms: Fruchterman-Reingold, Kamada-Kawai, and grid. Each of them has advantages and disadvantages.

(1) Fruchterman-Reingold dispalys the cells of individual predicted crypts close together with ample space between crypts. This makes it hard to see connections within individual crypts but allows to get a good overview of individual structures.

(2) Kamada-Kawai spaces individual cells well, so we can see individual connections between them. It may, however, happen that two different predicted crypts are partially overlayed, as can be seen for dark red and bright green lineages on the right side of the plot.

(3) When the network is layed out as a grid, one gets generally a good overview of the predicted lineages and their connections, but it's hard to see which connections belongs to which cell in the same row.

```{r, eval=T, echo=T, fig.height=6, fig.width=6}

visualize_as_network(GEMLI_items_crypts, cutoff=70, display_orphan=F, max_edge_width=1, ground_truth=T, include_labels=F, layout_style="fr")
visualize_as_network(GEMLI_items_crypts, cutoff=70, display_orphan=F, max_edge_width=1, ground_truth=T, include_labels=F, layout_style="kk")
visualize_as_network(GEMLI_items_crypts, cutoff=70, display_orphan=F, max_edge_width=1, ground_truth=T, include_labels=F, layout_style="grid")

```

## Adding cell type information to the GEMLI items list

The cells of individual intestinal crypts can be assigned to different cell types. This information can be added to the GEMLI items list as 'cell_type' slot in the form of a dataframe with column 'cell.ID' and 'cell.type'.

```{r, eval=T, echo=T}

load('GEMLI_crypts_example_cell_type_annotation.RData')
GEMLI_items_crypts[['cell_type']] = Crypts_annotation

```

## Color prediction network visualization by cell type

The visualization of the lineage predictions can now be colored by the cell type annotation. This allows to see the composition of individual intestinal crypts.

```{r, eval=T, echo=T, fig.height=6, fig.width=6}

visualize_as_network(GEMLI_items_crypts, cutoff=70, max_edge_width=5, display_orphan=F, include_labels=F, ground_truth=T, highlight_FPs=T, layout_style="kk", cell_type_colors=T)

```

Specific colors can be assigned to specific cell types by adding a dataframe with column 'cell.type' and 'color' in the GEMLI items list slot 'cell_type_color'. This can also allow to highlight just one or two selected cell types.

```{r, eval=T, echo=T, fig.height=6, fig.width=6}

# Adding custom color as cell_type_color element to GEMLI_items
cell.type <- unique(GEMLI_items_crypts[['cell_type']]$cell.type)
color <- c("#5386BD", "skyblue1", "darkgreen", "gold", "red", "darkred", "black")
Cell_type_color <- data.frame(cell.type, color)
GEMLI_items_crypts[['cell_type_color']] = Cell_type_color

# Make a visualization network
visualize_as_network(GEMLI_items_crypts, cutoff=70, max_edge_width=5, display_orphan=F, include_labels=F, ground_truth=T, highlight_FPs=T, layout_style="kk", cell_type_colors=T)

```
